src/file_io.rs

```diff
@@ -0,0 +1,119 @@
+use std::{fs, io, iter::Iterator};
+
+pub trait FsReadWriter {
+    fn read_dir(&self, path: &str) -> io::Result<fs::ReadDir>;
+}
+
+pub struct Fs {}
+
+impl FsReadWriter for Fs {
+    fn read_dir(&self, path: &str) -> io::Result<fs::ReadDir> {
+        fs::read_dir(path)
+    }
+}
+
```

Comment:
ok I think I see - it makes it possible to create a default implementation of `FsFilterer` which uses the std fs library?

Reply:
I just added a lot more. The purpose is to interface out access to the underlying fs module. Initial motivation was that I can't mock out the fs calls because it's all concrete types, so I made traits with the methods I needed and then added pass-through implementations which are used by default in running code. In tests though, I can make mock implementations of those traits and inject them in instead of the fs module.

I really like this pattern in general, I use it on all my data types. More than just tests, when everything is interfaced, it makes it really easy to then remove or rewrite parts of it without affecting other code.

It is more verbose though. Let me know what you think.

Reply:
Yea makes total sense!